x/protocolpool/keeper/keeper.go

```diff
@@ -90,6 +101,143 @@ func (k Keeper) GetCommunityPool(ctx context.Context) (sdk.Coins, error) {
+		}
+		err = k.ToDistribute.Set(ctx, toDistribute.Sub(fundsAllocated))
+		if err != nil {
+			return sdk.Coin{}, err
+		}
+
+		// set claimable to zero
+		err = k.RecipientFundDistribution.Set(ctx, recipient, math.ZeroInt())
+		if err != nil {
+			return sdk.Coin{}, err
+		}
+	} else {
+		curBal, err := k.RecipientFundDistribution.Get(ctx, recipient)
+		if err != nil {
+			return sdk.Coin{}, err
+		}
+		// add allocated fund amount to claimable balance
+		err = k.RecipientFundDistribution.Set(ctx, recipient, curBal.Add(fundsAllocated))
+		if err != nil {
+			return sdk.Coin{}, err
+		}
+	}
+
+	return withdrawnAmount, nil
+}
+
+func (k Keeper) SetToDistribute(ctx context.Context, amount sdk.Coins) error {
+	err := k.ToDistribute.Set(ctx, amount.AmountOf(sdk.DefaultBondDenom))
+	if err != nil {
+		return fmt.Errorf("error while setting ToDistribute: %v", err)
+	}
+	return nil
+}
+
+func (k Keeper) iterateAndUpdateFundsDistribution(ctx context.Context, toDistributeAmount math.Int) (string, error) {
+	totalPercentageToBeDistributed := math.ZeroInt()
+	err := k.RecipientFundPercentage.Walk(ctx, nil, func(key sdk.AccAddress, value math.Int) (stop bool, err error) {
+		totalPercentageToBeDistributed = totalPercentageToBeDistributed.Add(value)
+		return false, nil
+	})
+	if err != nil {
+		return "", err
+	}
+	if totalPercentageToBeDistributed.GT(math.NewInt(100)) {
+		return "", fmt.Errorf("total funds percentage cannot exceed 100")
+	}
+
+	poolMAcc := k.authKeeper.GetModuleAccount(ctx, types.ModuleName)
+	poolBal := k.bankKeeper.GetAllBalances(ctx, poolMAcc.GetAddress())
+	denom := poolBal.GetDenomByIndex(0)
+
+	toDistributeDec := sdk.NewDecCoins(sdk.NewDecCoin(denom, toDistributeAmount))
+
+	// Calculate the funds to be distributed based on the total percentage to be distributed
+	totalAmountToBeDistributed := toDistributeDec.MulDec(math.LegacyNewDecFromIntWithPrec(totalPercentageToBeDistributed, 2))
+	totalDistrAmount := totalAmountToBeDistributed.AmountOf(denom)
+
+	err = k.RecipientFundPercentage.Walk(ctx, nil, func(key sdk.AccAddress, value math.Int) (stop bool, err error) {
+		// Calculate the funds to be distributed based on the percentage
+		decValue := math.LegacyNewDecFromIntWithPrec(value, 2)
+		percentage := math.LegacyNewDecFromIntWithPrec(totalPercentageToBeDistributed, 2)
+		recipientAmount := totalDistrAmount.Mul(decValue).Quo(percentage)
+		recipientCoins := recipientAmount.TruncateInt()
+
+		// Set funds to be claimed
+		toClaim, err := k.RecipientFundDistribution.Get(ctx, key)
+		if err != nil {
+			return false, err
+		}
+		amount := toClaim.Add(recipientCoins)
+		err = k.RecipientFundDistribution.Set(ctx, key, amount)
+		if err != nil {
+			return false, err
+		}
+
+		// Subtract the coins to be distributed from toDistribute
+		toDistribute, err := k.ToDistribute.Get(ctx)
+		if err != nil {
+			return false, err
+		}
+		err = k.ToDistribute.Set(ctx, toDistribute.Sub(recipientCoins))
+		if err != nil {
+			return false, err
+		}
```

Comment:
this should be 0 at the end of the loop, why not set it once instead of n amount of times?

Reply:
no, toDistribute can have more amount and requested distributed percentages maynot be equal to 100 all the times. 
I don't think this should be 0.

Reply:
after we iterate how would this be 0? the rest of the funds are part of the community pool, they should be distributed. I think this design could drain the community pool. 

if todistribute has 100, 40 is sent to recipient funds, the other 60 is part of the community pool. Todistribute is now 0. 

in the current design we have 100 on todistribute, 40 is sent out to the recipient funds, 60 is still present on todistribute, making it so that the next person to call this would distribute 24 to recipient funds. Is that right?